File: Assets/Scripts/Player.cs
using System;
using System.Collections;
using System.Collections.Generic;
using Unity.VisualScripting;
using UnityEditor;
using UnityEngine;

public class Player : MonoBehaviour
{

    public GameObject PlayerParentObject { get; set; }
    public GameObject PlayerObject { get; set; }
    public bool FirstPlayer { get; set; }
    public Pillar Position { get; set; }
    public int X { get; set; }
    public int Z { get; set; }
    public Bag Bag { get; set; }
    public int XP { get; set; }
    public int Coins { get; set; }
    public int Energy { get; set; }
    public string Name { get; set; }
    public int DazeTurns { get; set; } = 0;
    public int FrozenTurns { get; set; } = 0;
    public int IncreasedBackpackTurns { get; set; } = 0;

    public void SetupPlayer(string name)
    {
        XP = int.Parse(PlayerPrefs.GetString("start_xp"));
        Coins = int.Parse(PlayerPrefs.GetString("start_coins"));
        Energy = int.Parse(PlayerPrefs.GetString("start_energy"));
        Name = name;
        Bag = gameObject.AddComponent<Bag>();
        Bag.Capacity = int.Parse(PlayerPrefs.GetString("backpack_default_storage_capacity"));
    }

    public string GetStats()
    {
        string stats = $@"
{Name}
Energy: {Energy}
XP: {XP}
Coins: {Coins}
Position: ({X}, {Z})
Increased backpack capacity duration: {IncreasedBackpackTurns}
Daze turns: {DazeTurns}
Frozen turns: {FrozenTurns}
Backpack capacity: {Bag.GetWeight() / Bag.Capacity}
Raw cheap crystal -> Count: {Bag.GetCountCheapRaw()}
Processed cheap crystal -> Count: {Bag.GetCountCheapProcessed()}
Raw expensive crystal -> Count: {Bag.GetCountExpensiveRaw()}
Processed expensive crystal -> Count: {Bag.GetCountExpensiveProcessed()}";
        return stats;
    }

    void Start()
    {

    }


    void Update()
    {

    }


    void OnMouseOver()
    {
        if (Input.GetMouseButtonDown(1))
        {
            Actions.BuildHouse(this.X, this.Z);
        }
    }

    public void SetPosition(Pillar pillar)
    {
        Position = pillar;
        X = Position.X;
        Z = Position.Z;
    }

    public void DecreaseEnergy(int amount)
    {
        Energy -= amount;
        if (Energy < 0)
            Energy = 0;
    }

    public void IncreaseEnergy(int amount)
    {
        Energy += amount;
        if (Energy > int.Parse(PlayerPrefs.GetString("max_energy")))
            Energy = int.Parse(PlayerPrefs.GetString("max_energy"));
    }

    public void InvalidMoveTakeEnergy()
    {
        Energy -= int.Parse(PlayerPrefs.GetString("invalid_turn_energy_penalty"));
    }

    public void TakeCoins(int c)
    {
        Coins -= c;
    }

    public void AddCoins(int c)
    {
        Coins += c;
    }

    public void AddXP(int xp)
    {
        XP += xp;
    }

    public bool IsFrozen()
    {
        return FrozenTurns > 0;
    }
    public bool IsDazed()
    {
        return DazeTurns > 0;
    }

    public void DecreaseDazeTurns()
    {
        if (DazeTurns > 0)
        {
            DazeTurns -= 1;
            if (!IsDazed())
            {
                HandleDaze3D(false);

            }
        }
    }

    public void AddDazeTurns()
    {
        if (!IsDazed())
        {
            HandleDaze3D(true);

        }
        DazeTurns += int.Parse(PlayerPrefs.GetString("number_of_daze_turns"));
    }

    private void HandleDaze3D(bool active)
    {

        Transform particles = PlayerObject.transform.GetChild(2);
        particles.gameObject.SetActive(active);
    }

    public void AddFrozenTurns()
    {
        if (!IsFrozen())
        {
            HandleFreeze3D(true);

        }
        FrozenTurns += int.Parse(PlayerPrefs.GetString("number_of_frozen_turns"));
    }

    private void HandleFreeze3D(bool active)
    {
        Transform firstChildTransform = PlayerObject.transform.GetChild(0);
        var icecube = firstChildTransform.GetChild(1);
        icecube.gameObject.SetActive(active);
        Transform particles = PlayerObject.transform.GetChild(1);
        particles.gameObject.SetActive(active);
    }

    public void DecreaseFrozenTurns()
    {
        if (FrozenTurns > 0)
        {
            FrozenTurns -= 1;
            if (!IsFrozen())
            {
                HandleFreeze3D(false);
            }
        }
    }
    public void AddIncreasedBackpackStorageTurns()
    {
        IncreasedBackpackTurns += int.Parse(PlayerPrefs.GetString("number_of_bigger_backpack_turns"));
        Bag.IncreaseBagCapacity();
    }
    public void DecreaseIncreasedBackpackStorageTurns()
    {
        if (IncreasedBackpackTurns > 0)
        {
            IncreasedBackpackTurns -= 1;
        }
        else
        {
            Bag.ResetBagCapacity();
            while (Bag.GetWeight() > int.Parse(PlayerPrefs.GetString("backpack_default_storage_capacity")))
            {
                Bag.RemoveLastInsertedCrystal();
            }
        }
    }


}

File: PythonConsoleApp/player.py
from colorama import Fore, Style

from constants import Constants


class Player:
    def __init__(self, name, x, y):
        self.name = name
        self.energy = Constants.start_energy
        self.xp = Constants.start_xp
        self.coins = Constants.start_coins
        self.x = x
        self.y = y
        self.x_base = x
        self.y_base = y
        self.frozen_turns = 0
        self.daze_turns = 0
        self.increased_backpack_turns = 0
        self.backpack_storage_capacity = Constants.backpack_default_storage_capacity
        self.crystals = []

    def is_in_base(self):
        return self.x_base == self.x and self.y_base == self.y

    def remove_expensive_crystals(self, number_of_crystals_to_remove):
        expensive_crystals = [crystal for crystal in self.crystals if crystal.is_expensive]
        if len(expensive_crystals) >= number_of_crystals_to_remove:
            removed_crystals = expensive_crystals[:number_of_crystals_to_remove]
            self.crystals = [crystal for crystal in self.crystals if crystal not in removed_crystals]

    def remove_cheap_crystals(self, number_of_crystals_to_remove):
        non_expensive_crystals = [crystal for crystal in self.crystals if not crystal.is_expensive]
        if len(non_expensive_crystals) >= number_of_crystals_to_remove:
            removed_crystals = non_expensive_crystals[:number_of_crystals_to_remove]
            self.crystals = [crystal for crystal in self.crystals if crystal not in removed_crystals]

    def decrease_energy(self, amount):
        self.energy -= amount
        if self.energy <= 0:
            self.energy = 0

    def increase_energy(self, amount):
        self.energy += amount
        if self.energy >= Constants.max_energy:
            self.energy = Constants.max_energy

    def add_coins(self, c):
        self.coins += c
        if self.coins > Constants.max_coins:
            self.coins = Constants.max_coins

    def remove_coins(self, c):
        self.coins -= c
        if self.coins < 0:
            self.coins = 0

    def is_frozen(self):
        return self.frozen_turns > 0

    def add_frozen_turns(self):
        self.frozen_turns += Constants.number_of_frozen_turns

    def decrease_frozen_turns(self):
        if self.frozen_turns > 0:
            self.frozen_turns -= - 1

    def is_dazed(self):
        return self.daze_turns > 0

    def decrease_daze_turns(self):
        if self.daze_turns > 0:
            self.daze_turns -= 1

    def add_daze_turns(self):
        self.daze_turns += Constants.number_of_daze_turns

    def add_increased_backpack_storage_turns(self):
        self.increased_backpack_turns += Constants.number_of_bigger_backpack_turns
        self.backpack_storage_capacity = (Constants.backpack_default_storage_capacity +
                                          Constants.increase_in_backpack_storage_capacity)

    def decrease_increased_backpack_storage_turns(self):
        if self.increased_backpack_turns > 0:
            self.increased_backpack_turns -= 1
        else:
            self.backpack_storage_capacity = Constants.backpack_default_storage_capacity
            if self.get_total_weight() > Constants.backpack_default_storage_capacity:
                while self.get_total_weight() > Constants.backpack_default_storage_capacity:
                    self.crystals.pop()

    def add_to_backpack_storage_(self, crystal) -> bool:
        if self.get_total_weight() + crystal.get_weight() <= self.backpack_storage_capacity:
            self.crystals.append(crystal)
            return True
        return False

    def get_total_weight(self):
        return sum(crystal.get_weight() for crystal in self.crystals)

    def get_total_weight_expensive_processed(self):
        return sum(crystal.get_weight() for crystal in self.crystals if crystal.is_expensive and crystal.is_processed)

    def get_total_weight_expensive_raw(self):
        return sum(
            crystal.get_weight() for crystal in self.crystals if crystal.is_expensive and not crystal.is_processed)

    def get_total_weight_cheap_processed(self):
        return sum(
            crystal.get_weight() for crystal in self.crystals if not crystal.is_expensive and crystal.is_processed)

    def get_total_weight_cheap_raw(self):
        return sum(
            crystal.get_weight() for crystal in self.crystals if not crystal.is_expensive and not crystal.is_processed)

    def get_count_expensive_processed(self):
        return sum(1 for crystal in self.crystals if crystal.is_expensive and crystal.is_processed)

    def get_count_expensive_raw(self):
        return sum(1 for crystal in self.crystals if crystal.is_expensive and not crystal.is_processed)

    def get_count_expensive(self):
        return sum(1 for crystal in self.crystals if crystal.is_expensive)

    def get_count_cheap(self):
        return sum(1 for crystal in self.crystals if not crystal.is_expensive)

    def get_count_cheap_processed(self):
        return sum(1 for crystal in self.crystals if not crystal.is_expensive and crystal.is_processed)

    def get_count_cheap_raw(self):
        return sum(1 for crystal in self.crystals if not crystal.is_expensive and not crystal.is_processed)

    def print_stats(self, color):
        print(
            f"{color}{self.name}{Style.RESET_ALL}\n"
            f"{Fore.GREEN}Energy: {self.energy}{Style.RESET_ALL} "
            f"{Fore.CYAN}XP: {self.xp}{Style.RESET_ALL} "
            f"{Fore.YELLOW}Coins: {self.coins}{Style.RESET_ALL} "
            f"{Fore.MAGENTA}Position: ({self.x},{self.y}){Style.RESET_ALL}\n"
            f"{Fore.LIGHTGREEN_EX}Increased backpack capacity duration: {self.increased_backpack_turns}{Style.RESET_ALL} "
            f"{Fore.LIGHTRED_EX}Daze turns: {self.daze_turns}{Style.RESET_ALL} "
            f"{Fore.LIGHTCYAN_EX}Frozen turns: {self.frozen_turns}{Style.RESET_ALL} "
            f"{Fore.LIGHTYELLOW_EX}Backpack capacity: {self.get_total_weight()}/{self.backpack_storage_capacity}{Style.RESET_ALL}\n"
            f"{Fore.YELLOW}Raw cheap crystal -> Weight: {self.get_total_weight_cheap_raw()}, Count: {self.get_count_cheap_raw()} || "
            f"{Fore.YELLOW}Processed cheap crystal -> Weight: {self.get_total_weight_cheap_processed()}, Count: {self.get_count_cheap_processed()}{Style.RESET_ALL}\n"
            f"{Fore.LIGHTMAGENTA_EX}Raw expensive crystal -> Weight: {self.get_total_weight_expensive_raw()}, Count: {self.get_count_expensive_raw()} || "
            f"{Fore.LIGHTMAGENTA_EX}Processed expensive crystal -> Weight: {self.get_total_weight_expensive_processed()}, Count: {self.get_count_expensive_processed()}{Style.RESET_ALL}\n"
        )

    def add_xp(self, xp):
        self.xp += xp